File: crates/netcfgd-model/src/dns.rs
//! DNS policy as a per-link scope, and what each delivery mode can express.

use serde::{Deserialize, Serialize};
use std::net::IpAddr;

/// How a resolved DNS policy reaches the system.
///
/// A mode is a contract with a specific tool, not a preference. There is
/// deliberately no `Auto`: the mode decides where queries go, and that is not
/// something to pick by heuristic (`docs/decisions/0007`).
#[derive(Debug, Clone, PartialEq, Eq, Default, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum DnsMode {
	/// netcfgd does not manage DNS.
	#[default]
	None,
	/// Write `/etc/resolv.conf` directly.
	WriteResolvConf,
	/// Hand a flat per-interface blob to any `resolvconf` implementation.
	Resolvconf,
	/// Use openresolv's `private_interfaces` and subscriber mechanism, which
	/// carries scopes. The path to recommend: no systemd, and the same
	/// upstream as the dhcpcd this project already delegates leases to.
	Openresolv,
	/// Hand scopes to `systemd-resolved`. Pulls in systemd on any host that
	/// selects it, which is why it is one mode among several and never a
	/// default.
	Resolved,
	/// Write dnsmasq configuration directly.
	Dnsmasq,
	/// Write unbound configuration directly.
	Unbound,
	/// Hand the whole scoped structure to a script as JSON on stdin. The
	/// escape hatch; it receives scopes, not a flattened list.
	Exec(String),
}

impl DnsMode {
	/// Whether this mode can express per-domain query routing.
	///
	/// `resolv.conf` cannot: its `search` line is suffix completion, not
	/// routing, and there is no per-domain server concept in the format. A
	/// config that asks a flat mode for routing domains is an error rather
	/// than something to flatten, because flattening sends internal queries to
	/// a public resolver.
	#[must_use]
	pub fn can_route(&self) -> bool {
		match self {
			Self::None | Self::WriteResolvConf | Self::Resolvconf => false,
			Self::Openresolv | Self::Resolved | Self::Dnsmasq | Self::Unbound | Self::Exec(_) => {
				true
			}
		}
	}

	/// A stable name for diagnostics.
	#[must_use]
	pub fn name(&self) -> &'static str {
		match self {
			Self::None => "none",
			Self::WriteResolvConf => "write_resolv_conf",
			Self::Resolvconf => "resolvconf",
			Self::Openresolv => "openresolv",
			Self::Resolved => "resolved",
			Self::Dnsmasq => "dnsmasq",
			Self::Unbound => "unbound",
			Self::Exec(_) => "exec",
		}
	}
}

/// A single upstream resolver.
///
/// A bare address cannot carry a port or the name to validate a certificate
/// against, so this is a struct from the start even though `DoT` lands much
/// later -- widening it after the M4 freeze would be a major version bump.
#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct DnsServer {
	/// The resolver's address.
	pub addr: IpAddr,
	/// Non-default port.
	#[serde(skip_serializing_if = "Option::is_none", default)]
	pub port: Option<u16>,
	/// Name to expect in the server's certificate, for `DnsTransport::Tls`.
	#[serde(skip_serializing_if = "Option::is_none", default)]
	pub sni: Option<String>,
}

/// A suffix whose queries route to this scope's servers.
///
/// Kept separate from `search`, which is suffix completion. resolved spells
/// both as one list distinguished by a `~` prefix; that is an accident of its
/// config format and a reliable source of confusion, and only one of the two
/// is universally supported.
#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct RoutingDomain {
	/// The suffix, or `"."` for the catch-all scope.
	pub suffix: String,
	/// Whether this scope's servers are used *only* for this suffix.
	#[serde(default)]
	pub exclusive: bool,
}

/// DNSSEC validation posture, passed to a backend that implements it.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum Dnssec {
	/// Do not validate.
	#[default]
	No,
	/// Validate where the zone is signed; do not fail otherwise.
	Allow,
	/// Require validation.
	Yes,
}

/// Transport to the upstream resolver.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum DnsTransport {
	/// Port 53, cleartext.
	#[default]
	Plain,
	/// DNS over TLS.
	Tls,
	/// DNS over HTTPS.
	Https,
}

/// One DNS scope: an interface's, or the host-wide fallback.
///
/// A per-interface policy is a scope in its own right. It is never merged into
/// a single global server list at compile time, because doing so destroys the
/// per-link structure at the earliest and least recoverable moment
/// (`docs/decisions/0007`).
#[derive(Debug, Clone, PartialEq, Eq, Default, Serialize, Deserialize)]
#[serde(deny_unknown_fields, default)]
pub struct DnsPolicy {
	/// How this scope reaches the system.
	pub mode: DnsMode,
	/// Upstream resolvers for this scope.
	pub servers: Vec<DnsServer>,
	/// Suffix completion. Every mode supports it.
	pub search: Vec<String>,
	/// Query routing. Only scope-capable modes support it.
	pub domains: Vec<RoutingDomain>,
	/// Resolver options passed through verbatim.
	pub options: Vec<String>,
	/// DNSSEC posture, where the backend implements it.
	#[serde(skip_serializing_if = "Option::is_none")]
	pub dnssec: Option<Dnssec>,
	/// Transport, where the backend implements it.
	#[serde(skip_serializing_if = "Option::is_none")]
	pub transport: Option<DnsTransport>,
}

impl DnsPolicy {
	/// Whether this scope asks for anything its mode cannot deliver.
	#[must_use]
	pub fn needs_routing(&self) -> bool {
		!self.domains.is_empty()
	}

	/// Whether this scope has nothing to deliver.
	///
	/// `dns { }` with no servers, no search and no options is a block that asks
	/// for nothing, and a scope for it is an action that does nothing. The mode
	/// is not consulted: a mode alone delivers no configuration, it only says
	/// how configuration would be delivered if there were any.
	#[must_use]
	pub fn is_empty(&self) -> bool {
		self.servers.is_empty()
			&& self.search.is_empty()
			&& self.domains.is_empty()
			&& self.options.is_empty()
	}
}

/// Every DNS scope a document asks for, given what has been observed.
///
/// **One function, called by both the planner and the executor**, because they
/// have to agree and once did not. The planner learned that a modem helper's
/// report contributes nameservers (decision 0006 rule 4) and the executor kept
/// building its scope list from the document alone, so the plan said
/// `dns.apply` and the delivery wrote a `resolv.conf` with nothing in it. That
/// is the same failure `make executor-policy` exists to prevent one crate over,
/// and the same answer: not two implementations kept in step, one implementation.
///
/// Scopes are owned rather than borrowed because one of them does not exist in
/// the document -- it is synthesised from an observation, and there is nothing
/// to borrow it from.
///
/// The order is deterministic: globals first, then interfaces in document
/// order. `netcfgd-dns` reverses that when it flattens, on purpose -- a
/// specific answer is consulted before the fallback -- and that is its business
/// rather than this function's.
#[must_use]
pub fn scopes(document: &crate::Document, observed: &crate::Observed) -> Vec<(String, DnsPolicy)> {
	let mut scopes: Vec<(String, DnsPolicy)> = Vec::new();
	let global_mode = &document.globals.dns.mode;
	if *global_mode != DnsMode::None {
		scopes.push(("globals".to_owned(), document.globals.dns.clone()));
	}

	for interface in &document.interfaces {
		// An unmanaged device contributes no scope. `dns.apply` is host-wide,
		// so it names no interface and the planner's usual choke point does not
		// see it -- this is where it has to be asked.
		if document
			.devices
			.iter()
			.any(|device| device.name == interface.name && !device.managed)
		{
			continue;
		}

		let reported = reported_servers(interface, observed);
		match (&interface.dns, reported.is_empty()) {
			// A policy with a mode of its own and nothing to say. `dns { }`
			// with no servers, no search and no options is a block that asks
			// for nothing, and a scope for it is an action that does nothing.
			(Some(policy), true) if policy.is_empty() => {}
			(Some(policy), true) => {
				scopes.push((interface.name.clone(), inheriting(policy, global_mode)));
			}
			// Rule 4: a source contributes nameservers and they merge with what
			// the document wrote. The document's come first, so
			// first-occurrence-wins means a server an operator chose beats one
			// the network handed out.
			(Some(policy), false) => {
				let mut merged = inheriting(policy, global_mode);
				merged.servers.extend(reported);
				merged.servers.dedup();
				scopes.push((interface.name.clone(), merged));
			}
			// No `dns` block, but something reported servers. The mode is not a
			// choice: `netcfgd-dns` refuses a delivery whose scopes disagree
			// about it, so the only value that is not an error is the one the
			// rest of the host uses.
			(None, false) if *global_mode != DnsMode::None => {
				let mut policy = DnsPolicy {
					mode: global_mode.clone(),
					..DnsPolicy::default()
				};
				policy.servers.extend(reported);
				scopes.push((interface.name.clone(), policy));
			}
			// Servers reported and nowhere to put them, because this host
			// manages no DNS. Skipped rather than pushed with a `None` mode: a
			// scope that delivers nothing is still an action in the plan, and an
			// action that does nothing is one somebody reads and dismisses on
			// every run. A host that does not manage DNS should not start
			// because a modem appeared.
			(None, _) => {}
		}
	}
	scopes
}

/// A policy with the host's mode filled in where it states none of its own.
///
/// **The mode is not a per-interface choice.** `netcfgd-dns` refuses a delivery
/// whose scopes disagree about it, because a host cannot both own `resolv.conf`
/// and hand it to `resolvconf`. So `none` on a scope that has something to
/// deliver can only mean "not stated", and the only value that is not an error
/// is the one the rest of the host uses.
///
/// This closes a defect older than the modem work that surfaced while merging
/// two implementations of the scope list into one: `dns = "9.9.9.9"` on an
/// interface compiles to a policy with mode `none`, and the executor dropped
/// the scope -- so an operator wrote a nameserver down and netcfgd silently
/// ignored it. Nothing failed and nothing warned; the server simply never
/// reached `resolv.conf`.
fn inheriting(policy: &DnsPolicy, global_mode: &DnsMode) -> DnsPolicy {
	if policy.mode != DnsMode::None {
		return policy.clone();
	}
	DnsPolicy {
		mode: global_mode.clone(),
		..policy.clone()
	}
}

/// The nameservers an interface's sources contribute, from what was observed.
///
/// Only a `modem` source contributes any today -- a DHCP lease's servers belong
/// to the client that took the lease, which writes them itself. Gated on the
/// document actually asking for the source, because a report netcfgd has no
/// instruction about is an observation, and a resolver is not something to
/// configure off a file somebody dropped in `/run`.
fn reported_servers(interface: &crate::Interface, observed: &crate::Observed) -> Vec<DnsServer> {
	if !interface
		.addressing
		.iter()
		.any(|source| matches!(source, crate::AddressSource::Modem(_)))
	{
		return Vec::new();
	}
	observed
		.modems
		.iter()
		.filter(|modem| modem.interface == interface.name)
		.flat_map(|modem| modem.nameservers.iter())
		.filter_map(|server| {
			// Kept as text by the reader on purpose, so one bad line does not
			// discard a report. This is where it has to become an address.
			Some(DnsServer {
				addr: server.parse().ok()?,
				port: None,
				sni: None,
			})
		})
		.collect()
}

File: c/src/model/dns.c
/*
 * scopes.c -- which scopes a machine has, given a document and an observation.
 *
 * WHY THIS IS NOT IN THE PLANNER, WHERE IT WAS WRITTEN
 *   It arrived as a static in `plan/host_wide.c`, because the planner was the
 *   only thing that had ever needed it, and that file's header said what the
 *   condition for moving it was: a second caller takes this one rather than
 *   writing a third. The daemon is the second caller -- `ncfg_service_t` wants
 *   every scope so that one `dns.apply` delivers all of them -- and the Rust
 *   records what happens when the two lists are built separately: the planner
 *   learned that a lease contributes nameservers and the executor went on
 *   reading the document alone, so the plan said `dns.apply` and the delivery
 *   wrote a `resolv.conf` with nothing in it.
 *
 * WHY IT OWNS AN ARENA AND THE PLANNER'S VERSION DID NOT
 *   The old one interned its merged policies into the plan, which is where its
 *   one caller kept everything else. A rule with two callers cannot borrow one
 *   caller's arena, so it has its own -- a list of blocks and one loop, which
 *   is `ncfg_plan_t`'s arrangement and is here for its reason: a policy is a
 *   struct with two re-allocated lists hanging off it, and a free walk per
 *   scope is a second list to keep in step with the struct.
 *
 *   The consequence is the sentence in `dns.h`: what comes back must outlive
 *   anything holding a scope from it. A plan holds them -- its `dns.apply` ops
 *   borrow the policies -- so `ncfg_plan_dns` adopts the whole object rather
 *   than copying out of it.
 *
 * WHAT IS DELIBERATELY NOT HERE
 *   Whether a delivery is worth making. "Every scope would deliver nothing"
 *   produces a plan warning and a decision about what to write, and both are
 *   the planner's: this answers what the scopes *are*, and `host_wide.c`
 *   decides what to do about them. The split is the one that lets the daemon
 *   ask the question without inheriting an opinion about plans.
 */
#include "ncfg/observed.h"

#include "ncfg/base.h"
#include "ncfg/value.h"

#include <stdint.h>
#include <stdlib.h>
#include <string.h>

struct ncfg_dns_scopes {
	ncfg_dns_scope_t *items;
	size_t            count;
	size_t            capacity;
	/* Every block the merges allocated: the policy copies, their two grown
	 * lists, and the canonical spelling of each reported server. */
	void            **owned;
	size_t            owned_count;
	size_t            owned_capacity;
};

/* ------------------------------------------------------------------------ *
 * The arena
 * ------------------------------------------------------------------------ */

/* Record one allocation as the list's, or free it and answer NULL. */
static void *keep(ncfg_dns_scopes_t *scopes, void *block)
{
	void **grown;
	size_t wanted;

	if (!block) {
		return NULL;
	}
	if (scopes->owned_count == scopes->owned_capacity) {
		wanted = scopes->owned_capacity ? scopes->owned_capacity * 2u : 16u;
		grown = realloc(scopes->owned, wanted * sizeof(*grown));
		if (!grown) {
			free(block);
			return NULL;
		}
		scopes->owned = grown;
		scopes->owned_capacity = wanted;
	}
	scopes->owned[scopes->owned_count++] = block;
	return block;
}

static const char *intern(ncfg_dns_scopes_t *scopes, const char *text)
{
	size_t length;
	char  *copy;

	if (!text) {
		return NULL;
	}
	length = strlen(text);
	copy = malloc(length + 1u);
	if (copy) {
		memcpy(copy, text, length + 1u);
	}
	return keep(scopes, copy);
}

static int scopes_push(ncfg_dns_scopes_t *scopes, const char *name,
    const ncfg_dns_policy_t *policy)
{
	ncfg_dns_scope_t *grown;
	size_t            wanted;

	if (!policy) {
		return 0;
	}
	if (scopes->count == scopes->capacity) {
		wanted = scopes->capacity ? scopes->capacity * 2u : 8u;
		grown = realloc(scopes->items, wanted * sizeof(*grown));
		if (!grown) {
			return 0;
		}
		scopes->items = grown;
		scopes->capacity = wanted;
	}
	scopes->items[scopes->count].name = name;
	scopes->items[scopes->count].policy = policy;
	scopes->count++;
	return 1;
}

/* ------------------------------------------------------------------------ *
 * What a report contributes
 * ------------------------------------------------------------------------ */

/*
 * Whether this interface asked for what a report has to offer.
 *
 * **The same gate for the servers and the suffixes, and that is the argument
 * rather than a convenience**: a suffix is only delivered where that network's
 * resolvers are already answering, and a party answering every query gains
 * nothing by also appending a suffix. Where an operator kept their own
 * resolvers, a lease does not get to redefine what a bare name means -- which
 * is the case 0049 was protecting.
 */
static int claimed(const ncfg_interface_t *interface)
{
	size_t i;

	if (interface->dns) {
		return 1;
	}
	for (i = 0; i < interface->addressing_count; i++) {
		if (interface->addressing[i].kind == NCFG_ADDRESS_SOURCE_REPORTED) {
			return 1;
		}
	}
	return 0;
}

static const ncfg_observed_report_t *report_for(const ncfg_observed_t *observed, const char *name)
{
	size_t i;

	if (!observed) {
		return NULL;
	}
	for (i = 0; i < observed->report_count; i++) {
		if (observed->reports[i].interface && name &&
		    strcmp(observed->reports[i].interface, name) == 0) {
			return &observed->reports[i];
		}
	}
	return NULL;
}

/*
 * A nameserver a report named, as the model spells it.
 *
 * Kept as text by the reader on purpose, so one bad line does not discard a
 * whole report; this is where it has to become an address. A prefix is
 * refused rather than carried, because a nameserver is one address and
 * `10.0.0.53/24` in a `nameserver` line is a resolver file nothing can use.
 *
 * **Canonical, and that is what makes the pass idempotent.** What an apply
 * delivered is folded into `owned.json` and comes back as `observed.dns`, and
 * the planner compares the two: a server carried through in whatever spelling
 * the report's author used would compare unequal against netcfgd's own record
 * of having delivered it, and `dns.apply` would be planned on every run.
 */
static const char *reported_server(ncfg_dns_scopes_t *scopes, const char *text)
{
	ncfg_address_t parsed;
	char           canonical[NCFG_ADDRESS_MAX];

	if (!text || !ncfg_address_parse(text, &parsed, NULL, 0) || parsed.has_prefix) {
		return NULL;
	}
	if (!ncfg_address_render(&parsed, canonical, sizeof(canonical), NULL, 0)) {
		return NULL;
	}
	return intern(scopes, canonical);
}

/* ------------------------------------------------------------------------ *
 * Building one scope
 * ------------------------------------------------------------------------ */

/*
 * A copy of `policy` with room for what a report is about to add.
 *
 * Only the two lists a merge appends to are re-allocated; everything else stays
 * exactly the borrow it was. A scope whose servers came straight off the
 * document never reaches here at all -- `build_scope` hands the document's
 * policy back -- so the copy exists precisely because these two lists are about
 * to be something neither the document nor the observation holds.
 */
static ncfg_dns_policy_t *policy_copy(ncfg_dns_scopes_t *scopes, const ncfg_dns_policy_t *policy,
    size_t extra_servers, size_t extra_search)
{
	ncfg_dns_policy_t *copy = calloc(1u, sizeof(*copy));
	size_t             servers;
	size_t             search;

	if (!keep(scopes, copy)) {
		return NULL;
	}
	if (policy) {
		*copy = *policy;
	}
	servers = copy->server_count + extra_servers;
	search = copy->search_count + extra_search;
	if (servers != 0u) {
		ncfg_dns_server_t *room = calloc(servers, sizeof(*room));

		if (!keep(scopes, room)) {
			return NULL;
		}
		if (copy->server_count != 0u) {
			memcpy(room, copy->servers, copy->server_count * sizeof(*room));
		}
		copy->servers = room;
	}
	if (search != 0u) {
		char **room = calloc(search, sizeof(*room));

		if (!keep(scopes, room)) {
			return NULL;
		}
		if (copy->search_count != 0u) {
			memcpy(room, copy->search, copy->search_count * sizeof(*room));
		}
		copy->search = room;
	}
	return copy;
}

/*
 * A policy with the host's mode filled in where it states none of its own.
 *
 * **The mode is not a per-interface choice.** The delivery refuses a set of
 * scopes that disagree about it, because a host cannot both own `resolv.conf`
 * and hand it to `resolvconf`. So `none` on a scope that has something to
 * deliver can only mean "not stated", and the only value that is not an error
 * is the one the rest of the host uses. Without this, `dns = "9.9.9.9"` on an
 * interface compiles to a policy with mode `none`, the executor drops the
 * scope, and an operator wrote a nameserver down that netcfgd silently
 * ignored.
 *
 * Returns the document's own policy where it needs nothing added, so the
 * ordinary case borrows exactly as `dns.h` says a scope does.
 */
static const ncfg_dns_policy_t *build_scope(ncfg_dns_scopes_t *scopes,
    const ncfg_dns_mode_value_t *host, const ncfg_dns_policy_t *base,
    const ncfg_observed_report_t *report)
{
	ncfg_dns_policy_t *copy;
	size_t             extra_servers = report ? report->nameserver_count : 0u;
	size_t             extra_search = report ? report->search_count : 0u;
	size_t             i;

	if (base && base->mode.mode != NCFG_DNS_MODE_NONE && !extra_servers && !extra_search) {
		return base;
	}
	copy = policy_copy(scopes, base, extra_servers, extra_search);
	if (!copy) {
		return NULL;
	}
	if (!base || base->mode.mode == NCFG_DNS_MODE_NONE) {
		copy->mode = *host;
	}
	/*
	 * Rule 4: a source contributes nameservers and they merge with what the
	 * document wrote. **The document's come first**, so first-occurrence-wins
	 * means a server an operator chose beats one the network handed out -- and
	 * the same for the suffixes, which resolution tries in order.
	 */
	for (i = 0; report && i < report->nameserver_count; i++) {
		const char *address = reported_server(scopes, report->nameservers[i]);

		if (!address) {
			continue;
		}
		/* Consecutive duplicates only, which is what the Rust's `dedup` does:
		 * a report repeating what the document already ends with contributes
		 * nothing, and two identical servers an operator wrote on purpose are
		 * left alone. */
		if (copy->server_count != 0u && copy->servers[copy->server_count - 1u].addr &&
		    strcmp(copy->servers[copy->server_count - 1u].addr, address) == 0) {
			continue;
		}
		memset(&copy->servers[copy->server_count], 0, sizeof(*copy->servers));
		copy->servers[copy->server_count].addr = (char *)(uintptr_t)address;
		copy->server_count++;
	}
	for (i = 0; report && i < report->search_count; i++) {
		if (!report->search[i]) {
			continue;
		}
		if (copy->search_count != 0u && copy->search[copy->search_count - 1u] &&
		    strcmp(copy->search[copy->search_count - 1u], report->search[i]) == 0) {
			continue;
		}
		copy->search[copy->search_count++] = report->search[i];
	}
	return copy;
}

static int policy_is_empty(const ncfg_dns_policy_t *policy)
{
	return policy->server_count == 0u && policy->search_count == 0u &&
	    policy->domain_count == 0u && policy->option_count == 0u;
}

/* ------------------------------------------------------------------------ *
 * The rule
 * ------------------------------------------------------------------------ */

ncfg_dns_scopes_t *ncfg_dns_scopes_of(const ncfg_document_t *desired,
    const ncfg_observed_t *observed, char *err, size_t err_size)
{
	ncfg_dns_scopes_t *scopes;
	size_t             i;

	if (!desired) {
		ncfg_error_set(err, err_size, "there is no document to take DNS scopes from");
		return NULL;
	}
	scopes = calloc(1u, sizeof(*scopes));
	if (!scopes) {
		ncfg_error_set(err, err_size, "out of memory taking the DNS scopes");
		return NULL;
	}
	if (desired->globals.dns.mode.mode != NCFG_DNS_MODE_NONE) {
		if (!scopes_push(scopes, NCFG_DNS_GLOBAL_SCOPE, &desired->globals.dns)) {
			goto no_memory;
		}
	}
	for (i = 0; i < desired->interface_count; i++) {
		const ncfg_interface_t       *interface = &desired->interfaces[i];
		const ncfg_device_t          *device = ncfg_document_device(desired, interface->name);
		const ncfg_observed_report_t *report;
		const ncfg_dns_policy_t      *policy;
		int                           offers = 0;

		/*
		 * An unmanaged device contributes no scope. `dns.apply` is host-wide,
		 * so it names no interface and the planner's usual choke point does
		 * not see it -- this is where it has to be asked.
		 */
		if (device && !device->managed) {
			continue;
		}
		report = claimed(interface) ? report_for(observed, interface->name) : NULL;
		offers = report && (report->nameserver_count != 0u || report->search_count != 0u);
		if (!interface->dns) {
			/*
			 * Servers reported and nowhere to put them, because this host
			 * manages no DNS. Skipped rather than pushed with a `none` mode: a
			 * scope that delivers nothing is still an action in a plan, and an
			 * action that does nothing is one somebody reads and dismisses on
			 * every run.
			 */
			if (!offers || desired->globals.dns.mode.mode == NCFG_DNS_MODE_NONE) {
				continue;
			}
		} else if (!offers && policy_is_empty(interface->dns)) {
			/* `dns { }` with nothing in it asks for nothing, and a scope for
			 * it is an action that does nothing. */
			continue;
		}
		policy = build_scope(scopes, &desired->globals.dns.mode, interface->dns,
		    offers ? report : NULL);
		if (!policy || !scopes_push(scopes, interface->name, policy)) {
			goto no_memory;
		}
	}
	return scopes;

no_memory:
	ncfg_dns_scopes_free(scopes);
	ncfg_error_set(err, err_size, "out of memory taking the DNS scopes");
	return NULL;
}

const ncfg_dns_scope_t *ncfg_dns_scopes_items(const ncfg_dns_scopes_t *scopes, size_t *count)
{
	static const ncfg_dns_scope_t none;

	if (count) {
		*count = scopes ? scopes->count : 0u;
	}
	if (!scopes) {
		return NULL;
	}
	/*
	 * An empty list answers a pointer rather than NULL, which `dns.h`
	 * promises: a caller that took NULL for "could not be asked" would
	 * otherwise read a document managing no DNS as a failure. `items` is NULL
	 * until the first push, so the answer comes from a value with a place in
	 * memory and a count of zero beside it.
	 */
	return scopes->items ? scopes->items : &none;
}

void ncfg_dns_scopes_free(ncfg_dns_scopes_t *scopes)
{
	size_t i;

	if (!scopes) {
		return;
	}
	for (i = 0; i < scopes->owned_count; i++) {
		free(scopes->owned[i]);
	}
	free(scopes->owned);
	free(scopes->items);
	free(scopes);
}
